tests/pytests/functional/modules/test_service.py

```diff
@@ -0,0 +1,157 @@
+import os
+
+import pytest
+
+import salt.utils.path
+import salt.utils.platform
+import salt.utils.systemd
+from salt.exceptions import CommandExecutionError
+
+pytestmark = [
+    pytest.mark.windows_whitelisted,
+    pytest.mark.destructive_test,
+    pytest.mark.slow_test,
+]
+
+
+@pytest.fixture
+def service_name(grains, modules):
+    # For local testing purposes
+    env_name = os.environ.get("SALT_FUNCTIONAL_TEST_SERVICE_NAME")
+    if env_name is not None:
+        return env_name
+
+    service_name = "cron"
+    cmd_name = "crontab"
+    os_family = grains.get("os_family")
+    is_systemd = grains.get("systemd")
+    if os_family == "RedHat":
+        service_name = "crond"
+    elif os_family == "Arch":
+        service_name = "sshd"
+        cmd_name = "systemctl"
+    elif os_family == "MacOS":
+        service_name = "com.apple.AirPlayXPCHelper"
+    elif os_family == "Windows":
+        service_name = "Spooler"
+
+    if os_family != "Windows" and salt.utils.path.which(cmd_name) is None:
+        pytest.skip("{} is not installed".format(cmd_name))
+
+    if is_systemd and modules.service.offline():
+        pytest.skip("systemd is OFFLINE")
+
+    return service_name
+
+
+@pytest.fixture(autouse=True)
+def setup_service(service_name, modules):
+    pre_srv_status = modules.service.status(service_name)
+    pre_srv_enabled = service_name in modules.service.get_enabled()
+
+    yield pre_srv_status
+
+    post_srv_status = modules.service.status(service_name)
+    post_srv_enabled = service_name in modules.service.get_enabled()
+
+    if post_srv_status != pre_srv_status:
+        if pre_srv_status:
+            modules.service.enable(service_name)
+        else:
+            modules.service.disable(service_name)
+
+    if post_srv_enabled != pre_srv_enabled:
+        if pre_srv_enabled:
+            modules.service.enable(service_name)
+        else:
+            modules.service.disable(service_name)
```

Comment:
```suggestion
def setup_service(service_name, modules):
    pre_srv_status = modules.service.status(service_name)
    pre_srv_enabled = service_name in modules.service.get_enabled()

    try:
        yield pre_srv_status
    finally:
        post_srv_status = modules.service.status(service_name)
        post_srv_enabled = service_name in modules.service.get_enabled()

        if post_srv_status != pre_srv_status:
            if pre_srv_status:
                modules.service.enable(service_name)
            else:
                modules.service.disable(service_name)

        if post_srv_enabled != pre_srv_enabled:
            if pre_srv_enabled:
                modules.service.enable(service_name)
            else:
                modules.service.disable(service_name)
```
We always want the `finally` block to run, even if tests fail.